File: Trigger/TrigT1/TrigT1CaloMonitoring/python/OverviewMonitorAlgorithm.py
#
#  Copyright (C) 2002-2022 CERN for the benefit of the ATLAS collaboration
#
def OverviewMonitoringConfig(inputFlags):
    '''Function to configure LVL1 Overview algorithm in the monitoring system.'''

    from AthenaConfiguration.Enums import Format

    # get the component factory - used for getting the algorithms
    from AthenaConfiguration.ComponentFactory import CompFactory
    from AthenaConfiguration.ComponentAccumulator import ComponentAccumulator
    result = ComponentAccumulator()

    # make the athena monitoring helper
    from AthenaMonitoring import AthMonitorCfgHelper
    helper = AthMonitorCfgHelper(inputFlags,'OverviewMonitoringCfg')

    # get any algorithms
    OverviewMonAlg = helper.addAlgorithm(CompFactory.OverviewMonitorAlgorithm,'OverviewMonAlg')

    # add any steering
    groupName = 'OverviewMonitor' # the monitoring group name is also used for the package name
    OverviewMonAlg.PackageName = groupName

    mainDir = 'L1Calo'

    # add monitoring algorithm to group, with group name and main directory 
    myGroup = helper.addGroup(OverviewMonAlg, groupName , mainDir)

    # flag for online - different duration options required
    isOnline=inputFlags.Trigger.Online.isPartition and inputFlags.Input.Format is Format.BS

    # number of processed events
    global_labels = ["Processed Events"]
    myGroup.defineHistogram('n_processed;l1calo_1d_NumberOfEvents',
                            title='Number of processed events',                            
                            type='TH1I',
                            path='Overview/',
                            xbins=1,xmin=0,xmax=1, xlabels=global_labels)


    # global overview
    trigPath = 'Overview/Errors/'
    NumberOfGlobalErrors=15
    globalStatus_xlabels = [
        "PPMDataStatus",
        "PPMDataError",
        "SubStatus",
        "Parity",
        "LinkDown",
        "Transmission",
        "Simulation",
        "CMXSubStatus",
        "CMXParity",
        "CMXTransmission",
        "CMXSimulation",
        "RODStatus",
        "RODMissing",
        "ROBStatus",
        "Unpacking"]

    globalStatus_ylabels = []
    for crate in range(14):
        cr = crate
        if cr >= 12:
            cr -= 12
        if cr >= 8:
            cr -= 8
        type = "PP " if (crate < 8) else "CP " if (crate < 12) else "JEP "
        globalStatus_ylabels.append(type)
    
    myGroup.defineHistogram('globalOverviewX,globalOverviewY;l1calo_2d_GlobalOverview',title='L1Calo Global Error Overview;;',
                            type='TH2I',
                            path=trigPath,
                            xbins=NumberOfGlobalErrors,xmin=0.,xmax=NumberOfGlobalErrors,
                            ybins=14,ymin=0.,ymax=14, xlabels=globalStatus_xlabels, ylabels=globalStatus_ylabels,
                            duration='' if isOnline else 'lb',
                            opt='kAlwaysCreate')

    if isOnline:
        myGroup.defineHistogram('globalOverviewX,globalOverviewY;l1calo_2d_GlobalOverviewRecent',
                                title='L1Calo Global Error Overview Last 10 LumiBlocks;;',
                                type='TH2I',
                                path=trigPath,
                                xbins=NumberOfGlobalErrors,xmin=0.,xmax=NumberOfGlobalErrors,
                                ybins=14,ymin=0.,ymax=14, xlabels=globalStatus_xlabels, ylabels=globalStatus_ylabels,
                                opt='kLBNHistoryDepth=10,kAlwaysCreate')


    myGroup.defineHistogram('lb_errors;l1calo_1d_ErrorsByLumiblock',
                            title='Events with Errors by Lumiblock;Lumi Block;Number of Events;',                            
                            type='TH1I',
                            path=trigPath,
                            xbins=3000,xmin=0,xmax=3000,
                            weight='n_lb_errors',
                            opt='kAlwaysCreate')

    
    acc = helper.result()
    result.merge(acc)
    return result


if __name__=='__main__':
    # set input file and config options
    from AthenaConfiguration.AllConfigFlags import ConfigFlags
    import glob

    #inputs = glob.glob('/eos/atlas/atlastier0/rucio/data18_13TeV/physics_Main/00357750/data18_13TeV.00357750.physics_Main.recon.ESD.f1072/data18_13TeV.00357750.physics_Main.recon.ESD.f1072._lb0117._SFO-1._0201.1')
    inputs = glob.glob('/eos/atlas/atlastier0/rucio/data18_13TeV/physics_Main/00354311/data18_13TeV.00354311.physics_Main.recon.ESD.f1129/data18_13TeV.00354311.physics_Main.recon.ESD.f1129._lb0013._SFO-8._0001.1')


    ConfigFlags.Input.Files = inputs
    ConfigFlags.Output.HISTFileName = 'ExampleMonitorOutput_LVL1.root'

    ConfigFlags.lock()
    ConfigFlags.dump() # print all the configs

    from AthenaCommon.AppMgr import ServiceMgr
    ServiceMgr.Dump = False

    from AthenaConfiguration.MainServicesConfig import MainServicesCfg 
    from AthenaPoolCnvSvc.PoolReadConfig import PoolReadCfg
    cfg = MainServicesCfg(ConfigFlags)
    cfg.merge(PoolReadCfg(ConfigFlags))

    OverviewMonitorCfg = OverviewMonitoringConfig(ConfigFlags)
    cfg.merge(OverviewMonitorCfg)

    # message level for algorithm
    OverviewMonitorCfg.getEventAlgo('OverviewMonAlg').OutputLevel = 2 # 1/2 INFO/DEBUG
    # options - print all details of algorithms, very short summary 
    cfg.printConfig(withDetails=False, summariseProps = True)

    nevents=-1
    cfg.run(nevents)

File: Trigger/TrigT1/TrigT1CaloMonitoring/src/OverviewMonitorAlgorithm.cxx
/*
  Copyright (C) 2002-2022 CERN for the benefit of the ATLAS collaboration
*/

#include "OverviewMonitorAlgorithm.h"

OverviewMonitorAlgorithm::OverviewMonitorAlgorithm( const std::string& name, ISvcLocator* pSvcLocator )
  : AthMonitorAlgorithm(name,pSvcLocator)
{
}

StatusCode OverviewMonitorAlgorithm::initialize() {

  ATH_MSG_DEBUG("OverviewMonitorAlgorith::initialize");
  ATH_MSG_DEBUG("Package Name "<< m_packageName);
  
  ATH_CHECK(m_cpmErrorLocation.initialize());
  ATH_CHECK(m_cpmMismatchLocation.initialize());
  ATH_CHECK(m_ppmSimBSMismatchLocation.initialize());


  return AthMonitorAlgorithm::initialize();
}

StatusCode OverviewMonitorAlgorithm::fillHistograms( const EventContext& ctx ) const {

  ATH_MSG_DEBUG("OverviewMonitorAlgorithm::fillHistograms");

  std::vector<std::reference_wrapper<Monitored::IMonitoredVariable>> variables;


  // Update Global overview plot

  const int ppmCrates = 8;
  const int cpmCrates = 4;

  Monitored::Scalar<int> globalOverviewX = Monitored::Scalar<int>("globalOverviewX", 0);
  Monitored::Scalar<int> globalOverviewY = Monitored::Scalar<int>("globalOverviewY", 0);

  int err_per_LB=0;

  // CPM and CPM CMX Error data
  {
    const auto* errTES = SG::get(m_cpmErrorLocation, ctx);

    if (!errTES || errTES->size() != size_t(cpmCrates)) {
      ATH_MSG_INFO("No CPM error vector of expected size");
    } else {
      for (int crate = 0; crate < cpmCrates; ++crate) {
	const int err = (*errTES)[crate];
	if (err == 0)
	  continue;
	const int cr = crate + ppmCrates;

	if ((err >> CPMStatus) & 0x1){
	  globalOverviewX=SubStatus;
	  globalOverviewY=cr;
	  fill(m_packageName,globalOverviewX,globalOverviewY);
	  err_per_LB+=1;
	}
	if (((err >> CPMEMParity) & 0x1) || ((err >> CPMHadParity) & 0x1)) {
	  globalOverviewX=Parity;
	  globalOverviewY=cr;
	  fill(m_packageName,globalOverviewX,globalOverviewY);
	  err_per_LB+=1;
	}
	if (((err >> CPMEMLink) & 0x1) || ((err >> CPMHadLink) & 0x1)) {
	  globalOverviewX=LinkDown;
	  globalOverviewY=cr;
	  fill(m_packageName,globalOverviewX,globalOverviewY);
	  err_per_LB+=1;
	}
	if ((err >> CMXCPTobParity) & 0x1) {
	  globalOverviewX=GbCMXParity;
	  globalOverviewY=cr;
	  fill(m_packageName,globalOverviewX,globalOverviewY);
	  err_per_LB+=1;
	}
	if ((err >> CMXCPSumParity) & 0x1) {
	  globalOverviewX=GbCMXParity;
	  globalOverviewY=cr;
	  fill(m_packageName,globalOverviewX,globalOverviewY);
	  err_per_LB+=1;
	}
	if ((err >> CMXCPStatus) & 0x1) {
	  globalOverviewX=CMXSubStatus;
	  globalOverviewY=cr;
	  fill(m_packageName,globalOverviewX,globalOverviewY);
	  err_per_LB+=1;
	}
      } // crates
    }
  } 

  // CPM and CMX Simulation Mismatch data
  { 
	const auto* errTES = SG::get(m_cpmMismatchLocation, ctx);

    if (!errTES || errTES->size() != size_t(cpmCrates)) {
      ATH_MSG_INFO("No CPM mismatch vector of expected size");
    } else {
      for (int crate = 0; crate < cpmCrates; ++crate) {
	const int err = (*errTES)[crate];
	if (err == 0)
	  continue;
	const int cr = crate + ppmCrates;
	if (((err >> EMTowerMismatch) & 0x1) || ((err >> HadTowerMismatch) & 0x1)) {
	  globalOverviewX=Transmission;
	  globalOverviewY=cr;
	  fill(m_packageName,globalOverviewX,globalOverviewY);
	  err_per_LB+=1;
	}
	if (((err >> EMRoIMismatch) & 0x1) || ((err >> TauRoIMismatch) & 0x1)) {
	  globalOverviewX=Simulation;
	  globalOverviewY=cr;
	  fill(m_packageName,globalOverviewX,globalOverviewY);
	  err_per_LB+=1;
	}
	if (((err >> LeftCMXTobMismatch) & 0x1) ||
	    ((err >> RightCMXTobMismatch) & 0x1) ||
	    ((err >> RemoteSumMismatch) & 0x1)) {
	  globalOverviewX=CMXTransmission;
	  globalOverviewY=cr;
	  fill(m_packageName,globalOverviewX,globalOverviewY);
	  err_per_LB+=1;
	}
	if (((err >> LocalSumMismatch) & 0x1) ||
	    ((err >> TotalSumMismatch) &
	     0x1)) {
	  globalOverviewX=CMXSimulation;
	  globalOverviewY=cr;
	  fill(m_packageName,globalOverviewX,globalOverviewY);
	  err_per_LB+=1;
	}
      } // crates
    }
  }


  // PPM Mismatch data
  { 
    const auto* errTES = SG::get(m_ppmSimBSMismatchLocation, ctx);
    if (!errTES || errTES->size() != size_t(ppmCrates)) {
      ATH_MSG_INFO("No PPM SimBS mismatch vector of expected size");
    } else {
      for (int crate = 0; crate < ppmCrates; ++crate) {
	const int err = (*errTES)[crate];
	if (err == 0)
	  continue;
	if (((err >> LUTMismatch) & 0x1)){
	  globalOverviewX=Simulation;
	  globalOverviewY=crate;
	  fill(m_packageName,globalOverviewX,globalOverviewY);
	  err_per_LB+=1;
	}
      }
    }
  }



  // errors per lumiblock and events processed
  auto lb = GetEventInfo(ctx)->lumiBlock();

  auto n_processed  = Monitored::Scalar("n_processed",0);
  auto lb_errors  = Monitored::Scalar("lb_errors",lb);
  auto n_lb_errors  = Monitored::Scalar("n_lb_errors",err_per_LB);

  variables.push_back(n_processed);
  if (err_per_LB>0) {
    variables.push_back(lb_errors);
    variables.push_back(n_lb_errors);
  }

  fill(m_packageName,variables);
  variables.clear();
  return StatusCode::SUCCESS;
}


